### PR TITLE
Predict total consumption by fuel type

### PR DESCRIPTION
**Pull request by mikivee**
1. Predicting by HVAC consumption by fuel type rather than by end use. See [description](https://www.notion.so/rewiringamerica/Building-towards-an-MVP-9cb748ec5ef647fba50a302fa7f152ca?pvs=4#975bcd6c5cbe4dd7b170ca65fc48b901).
2. Predicting total consumption for all upgrades (excluding standalone range and dryer upgrades). See [description](https://www.notion.so/rewiringamerica/Building-towards-an-MVP-9cb748ec5ef647fba50a302fa7f152ca?pvs=4#70bea834972b410e9e2a17c5463e44a3)

### CODE REVIEW

File: src/databricks/model_evaluation.py
Comment (mikivee) [inline src/databricks/model_evaluation.py:1]: Ugh, sorry this is gonna make this more annoying to review, but it looks like the diff was large enough that with the rename, it is treating this as a whole add/delete file situation 😢 .

File: src/databricks/model_training.py [hunk 1]
```diff
@@ -32,140 +34,70 @@
...
 
-# MAGIC %md ## Load Data
+# DBTITLE 1,Set experiment location
+# location to store the experiment runs if in production mode:
+# specifying this allows for models trained in notebook or job to be written to same place
+EXPERIMENT_LOCATION = "/Shared/surrogate_model/"
 
 # COMMAND ----------
 
-# DBTITLE 1,Data loading function
```
Comment (mikivee) [inline src/databricks/model_training.py:88]: moved to datagen.py

File: src/databricks/model_evaluation.py [hunk 2]
```diff
@@ -0,0 +1,650 @@
...
+        .when(F.col("heating_appliance_type") == "Shared Heating", F.lit("Shared Heating"))
+        .when(F.col("heating_fuel") == "None", F.lit("No Heating"))
+        .otherwise(F.col("heating_fuel")),
+    ).withColumn(
+        "ac_type",
+        F.when(F.col("ac_type") == "Shared Cooling", F.lit("Shared Cooling"))
+        .when(F.col("ac_type") == "None", F.lit("No Cooling"))
+        .otherwise(F.col("ac_type")),
+    )
+)
```
Comment (mikivee) [inline src/databricks/model_evaluation.py:255]: similar to before

Comment (mikivee) [inline src/databricks/model_evaluation.py:255]: very similar to before

File: src/databricks/model_evaluation.py [hunk 3]
```diff
@@ -0,0 +1,650 @@
...
+
+    Returns:
+    - double: The APE value, or None if actual or pred is None or
+              the actual value is within the epsilon range of zero.
+    """
+    if abs_error is None:
+        return None
+    if abs(actual) < eps:
+        return None
+    return abs(abs_error / actual) * 100
```
Comment (mikivee) [inline src/databricks/model_evaluation.py:283]: very similar to before

File: src/databricks/model_evaluation.py [hunk 4]
```diff
@@ -0,0 +1,650 @@
...
+        F.round(f(F.col(colname)), round_precision).alias(
+            f"{f.__name__}_{colname}")
+        for f in [F.median, F.mean]
+        for colname, round_precision in [
+            ("absolute_error", 0),
+            ("absolute_percentage_error", 1),
+            ("absolute_error_savings", 0),
+            ("absolute_percentage_error_savings", 1),
+        ]
+    ]
```
Comment (mikivee) [inline src/databricks/model_evaluation.py:358]: same as before

File: src/databricks/model_evaluation.py [hunk 5]
```diff
@@ -0,0 +1,650 @@
...
+# DBTITLE 1,Save metrics table
+if not DEBUG:
+# save the metrics table tagged with the model name and version number
+    (
+        metrics_by_upgrade_type_model.write.format("delta")
+        .mode("overwrite")
+        .option("overwriteSchema", "true")
+        .option("userMetadata", MODEL_RUN_NAME)
+        .saveAsTable("ml.surrogate_model.evaluation_metrics")
+    )
```
Comment (mikivee) [inline src/databricks/model_evaluation.py:428]: pretty much the same as before, but adding total_metrics_by_fuel df.

**Review by tarrenpeterson — APPROVED** [review]
This looks great to me - I know there was a lot of work happening behind the scenes to roll back that final layer of the layer cake related to the custom loss function but you made it look easy. I have a couple small questions but the thorough documentation preempted most questions

File: src/databricks/model_training.py [hunk 6]
```diff
@@ -274,21 +223,35 @@ def convert_feature_dataframe_to_dict(
...
+# DBTITLE 1,Train model
+# Train keras model and log the model with the Feature Engineering in UC. Note that right we are skipping registering the model in the UC-- this requires storing the signature, which for unclear reasons, is slowing down inference more than 10x.
 
-#Init FeatureEngineering client
+# Init FeatureEngineering client
 fe = FeatureEngineeringClient()
 
 # Set the activation function and numeric data type for the model's layers
-layer_params = {"activation": "leaky_relu", "dtype": np.float32}
+layer_params = {
```
[the remark this answers is not in the document]
Comment (mikivee) [reply src/databricks/model_training.py:235]: Good question. So before we had `layers.Dense(64,  activation = "leaky_relu")(bm)` which first applies the linear mapping and then the non-linear activation. But since we want to apply batchnorm in between the linear map and the activation, we have to instead not apply any non-linear activation at first (i.e., activation='linear' which is the default), then pass it thru a batchnorm layer and then pass it through the leakyrelu layer. There were two additional things that made this confusing that I have now fixed to make this more readable and easier to modify in the future:
1. when you have repeated "blocks" like this (e.g, conv-batchnorm-relu) the standard thing is to define a function for this block rather than just repeat it over and over. I was just being lazy. I have now done this. 
2. I was still using this pattern of manually applying the activation after the dense layer in the combine trunk, even tho I wasn't using batchnorm, so I could have just passed the activation name directly. I removed the activation from the arguments here since this is kind of stupid now, and then updated layers in combine trunk to use the simpler pattern.

File: src/databricks/model_training.py [hunk 7]
```diff
@@ -338,55 +301,25 @@ def convert_feature_dataframe_to_dict(
...
 # DBTITLE 1,Inspect predictions using logged model
 # evaluate the unregistered model we just logged and make sure everything runs
-if DEBUG: 
-    print(run_id)
-    pred_df = model.score_batch(test_data=test_data, run_id=run_id)
-    pred_df.display()
+print(run_id)
+# mlflow.pyfunc.get_model_dependencies(model_uri=sm.get_model_uri(run_id=run_id))
+# Load the model using its registered name and version/stage from the MLflow model registry
+model_loaded = mlflow.pyfunc.load_model(model_uri=sm.get_model_uri(run_id=run_id))
```
[the remark this answers is not in the document]
Comment (mikivee) [reply src/databricks/model_training.py:311]: yeah that is fair... I think the reason I put it there is because failed at inference time for like 1000 different reasons so far, and most of the time, it means I did something wrong in training and have to retrain the model. So I figured it would be good if it failed there just in case I don't run the eval script and I think "oh yay successful model run" and then don't realize I need to retrain. I can nest it in the DEBUG if you feel strongly.

File: src/databricks/model_evaluation.py [hunk 8]
```diff
@@ -0,0 +1,695 @@
...
+        "Type",
+    ],
+    columns=["Metric", "Model"],
+    values="value",
+).sort_values(["Upgrade ID", "Type"])
+
+# COMMAND ----------
+
+# DBTITLE 1,Display Results
+metrics_by_upgrade_type
```
[the remark this answers is not in the document]
Comment (mikivee) [reply src/databricks/model_evaluation.py:553]: My sense was that this was mostly just due to the fact that the savings are larger for 4 than 3, and so when using these relative measures, 4 looks better than 3. If you look at the error for just the consumption rather than savings  [here](https://docs.google.com/spreadsheets/d/1XO9a_NPOcqaZrsisJSCs5T3R2enVSPJ-eKI1Ilgf-H8/edit?gid=86053945#gid=86053945) (the right half), you'll see that they are quite similar. But there may be something deeper going on and I just don't realize it..

**Review by mikivee — COMMENTED** [review]
Thanks @tarrenpeterson! Super helpful and thorough comments as always. Gonna assign back to you to just look over the unresolved comments if you wanna take a look at those before I merge.

File: src/databricks/build_feature_store.py [hunk 9]
```diff
@@ -175,82 +248,296 @@ def extract_cooling_efficiency(cooling_efficiency: str) -> float:
...
+    fuel_and_type, tank_volume, efficiency = split_wh_name(name)
+
+    # Extract the water heater type
+    if "Tankless" in fuel_and_type:
+        specs["water_heater_type"] = "Instantaneous"
+        specs["water_heater_tank_volume_gal"] = 0
+    elif fuel_and_type == "Electric Heat Pump":
+        specs["water_heater_type"] = "Heat Pump"
+
+    # Set the efficiency and recovery efficiency
```
[the remark this answers is not in the document]
Comment (tarrenpeterson) [reply src/databricks/build_feature_store.py:477]: No the docstring is sufficient - sorry for not catching that on first pass

File: src/databricks/model_training.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tarrenpeterson) [reply src/databricks/model_training.py:235]: Ok that makes sense - I'd missed that detail about the ordering of steps here and that the batchnorm function needed to be added in between the linear mapping and non-linear activation. Thanks for adding the function in as well!

File: src/databricks/model_training.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tarrenpeterson) [reply src/databricks/model_training.py:311]: I don't feel strongly on this point so will defer to your judgement here

File: src/databricks/model_evaluation.py [hunk 10]
```diff
@@ -0,0 +1,695 @@
...
+
+# MAGIC %md ### Run Inference
+
+# COMMAND ----------
+
+# DBTITLE 1,Run inference
+# run inference: output a N x M matrix of predictions
+# where N is the number of rows in the input data table
+# and M is the number of target columns
+# takes ~20s on 10,0000 and ~2m on 100,0000 samples
```
[the remark this answers is not in the document]
Comment (tarrenpeterson) [reply src/databricks/model_evaluation.py:106]: I think it's a great place to start without doing any work on optimization yet

File: src/databricks/model_evaluation.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tarrenpeterson) [reply src/databricks/model_evaluation.py:553]: Ahh that's a good call - I didn't think of it from that point of view

File: src/databricks/model_training.py [hunk 11]
```diff
@@ -18,154 +18,88 @@
...
 # DBTITLE 1,Load data
-train_data, val_data, test_data = load_data(n_subset=100 if DEBUG else None)
+train_data, val_data, test_data = load_data(n_train=1000 if DEBUG else None)
 
 # COMMAND ----------
 
 # DBTITLE 1,Initialize train/val data generators
-train_gen = DataGenerator(train_data=train_data)
-val_gen = DataGenerator(train_data=val_data)
+train_gen = DataGenerator(train_data=train_data, batch_size=256)
```
[the remark this answers is not in the document]
Comment (tarrenpeterson) [reply src/databricks/model_training.py:101]: It would make sense to make it the new default to me but I don't think it's a big deal either way

File: src/databricks/datagen.py [hunk 12]
```diff
@@ -359,4 +395,88 @@ def on_epoch_end(self):
...
+    unique_building_ids = data.where(F.col("upgrade_id") == 0).select("building_id")
+
+    # Split the building_ids into train, validation, and test sets (may not exactly match passed proportions)
+    p_train = 1 - p_val - p_test
+    train_ids, val_ids, test_ids = unique_building_ids.randomSplit(
+        weights=[p_train, p_val, p_test], seed=seed
+    )
+
+    # if n_train or n_test are passed, get the fraction of the train or test subset that this represents
+    if n_train or n_test:
```
[the remark this answers is not in the document]
Comment (tarrenpeterson) [reply src/databricks/datagen.py:465]: Ahh I understand the first point and did not consider that when making the suggestion that you subset for `n_train` or `n_test` before the train/test split. But agree about the second point here and thank you for correcting with the ceiling!

File: src/databricks/model_training.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mikivee) [reply src/databricks/model_training.py:101]: ya u right. i'll update.

File: src/databricks/model_training.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mikivee) [reply src/databricks/model_training.py:311]: okay I'm going to keep for now
